src/Collections-Sequenceable-Tests/SortedCollectionTest.class.st

```diff
@@ -500,6 +500,14 @@ SortedCollectionTest >> testAddAll2 [
 	sorted2 do: [ :each | self assert: (sorted includes: each)]
 ]
 
+{ #category : #'tests - basic' }
+SortedCollectionTest >> testAddLast [
+
+	self
+		should: [ #( 10 9 8 ) asSortedCollection addLast: 11 ]
+		raise: ShouldNotImplement
```

Comment:
Do we also have a positive test? 

| ar |
ar := #( 10 9 8 ) asSortedCollection.
ar addLast: 11.
self assert: ar asArray equals: #( 11 10 9 8)

Reply:
Uh, there isn't really a "positive" case—SortedCollection now `shouldNotImplement` `addLast:`, no exceptions. This is how other Smalltalk dialects do it, and `addFirst:` is already disallowed in Pharo. I suspect the only reason `addLast:` is _not_ already disallowed is that it is used internally by the `select:`/`collect:` family, and `SortedCollection>>addLast:`. The bulk of the changes in this PR are untangling that to use a separate, private API (`addNoSort:`—I'm not attached to the name btw) so that `addLast:` can be safely made `shouldNotImplement`.

My question as far as tests was about all the `collect:`-family methods I had to change that used to return a `SortedCollection` which might or might not be sorted (due to the actions of a `collect:` block), and thus might error later when inserting or re-sorting, which now return an OrderedCollection, analogous to regular `collect:`. I feel like trying to write tests asserting that they all return an OrderedCollection would be a tremendous amount of duplication for little gain, and given the change to `addLast:`, the _existing_ tests would fail if those methods had not been updated (because they would encounter the `shouldNotImplement` error). But I wondered if anyone with more familiarity with the test suite had an idea of something I could do that _would_ be worthwhile.

Reply:
Yes, I also think the current PR is okay regarding tests. Cancelling `addLast:` has a positive test: it's cancelled and throws an exception :)

Though, I don't understand why `OrderedCollection` needs to implement `addNoSort:`?

Reply:
`OrderedCollection` implements `addNoSort:` so that methods that need that behavior on both Ordered and SortedCollection—anything with `select:` in the name—can have a common implementation. It is perhaps a little strange to have a method talking about not sorting on a class that doesn't perform any in the first place, but frankly those two are such a ubiquitous part of the standard library that I don't think anyone will have any trouble figuring it out. I chose to implement it this way—and similarly for `copyForTransform:`—despite feeling a little awkward about the names because it just saves _so much code_ that would otherwise be a near-exact copy.